internal/core/src/segcore/ConcurrentVector.h

```diff
@@ -84,6 +90,113 @@ class VectorBase {
     virtual void
     set_data_raw(ssize_t element_offset, const void* source, ssize_t element_count) = 0;
 
+    void
+    set_data_raw(ssize_t element_offset, ssize_t element_count, const DataArray* data, const FieldMeta& field_meta) {
+        if (field_meta.is_vector()) {
+            if (field_meta.get_data_type() == DataType::VECTOR_FLOAT) {
+                return set_data_raw(element_offset, data->vectors().float_vector().data().data(), element_count);
+            } else if (field_meta.get_data_type() == DataType::VECTOR_BINARY) {
+                return set_data_raw(element_offset, data->vectors().binary_vector().data(), element_count);
+            } else {
+                PanicInfo("unsupported");
+            }
+        }
+
+        switch (field_meta.get_data_type()) {
+            case DataType::BOOL: {
+                return set_data_raw(element_offset, data->scalars().bool_data().data().data(), element_count);
+            }
+            case DataType::INT8: {
+                auto src_data = data->scalars().int_data().data();
+                std::vector<int8_t> data_raw(src_data.size());
+                std::copy_n(src_data.data(), src_data.size(), data_raw.data());
+                return set_data_raw(element_offset, data_raw.data(), element_count);
+            }
+            case DataType::INT16: {
+                auto src_data = data->scalars().int_data().data();
+                std::vector<int16_t> data_raw(src_data.size());
+                std::copy_n(src_data.data(), src_data.size(), data_raw.data());
```

Comment:
why INT8/INT16 need copy_n ?

Reply:
protobuf does not support int8/int16 and does type promotion automatically